src/stream/from_fn.rs

```diff
@@ -0,0 +1,99 @@
+use std::marker::PhantomData;
+use std::pin::Pin;
+
+use crate::future::Future;
+use crate::stream::Stream;
+use crate::task::{Context, Poll};
+
+/// A stream that yeilds elements by calling a closure.
+///
+/// This stream is constructed by [`from_fn`] function.
+///
+/// [`from_fn`]: fn.from_fn.html
+#[derive(Debug)]
+pub struct FromFn<F, Fut, T> {
+    f: F,
+    future: Option<Fut>,
+    __t: PhantomData<T>,
+}
+
+/// Creates a new stream where to produce each new element a provided closure is called.
+///
+/// This allows creating a custom stream with any behaviour without using the more verbose
+/// syntax of creating a dedicated type and implementing a `Stream` trait for it.
+///
+/// # Examples
+///
+/// ```
+/// # fn main() { async_std::task::block_on(async {
+/// #
+/// use async_std::prelude::*;
+/// use std::sync::{Mutex, Arc};
+/// use async_std::stream;
+///
+/// let count = Arc::new(Mutex::new(0u8));
+/// let s = stream::from_fn(|| {
+///     let count = Arc::clone(&count);
+///
+///     async move {
+///         *count.lock().unwrap() += 1;
+///
+///         if *count.lock().unwrap() > 3 {
+///             None
+///         } else {
+///             Some(*count.lock().unwrap())
+///         }
+///     }
+/// });
```

Comment:
I like how this example somewhat creatively works around the limitations with async closures right now. This is the kind of thing people may want to do for sure, and I think there's value in showing how to do it _now_ also.